language/vm/vm_runtime/vm_runtime_types/src/value.rs

```diff
@@ -184,228 +126,471 @@ pub struct RootAccessPath {
 }
 
-impl Clone for MutVal {
-    fn clone(&self) -> Self {
-        MutVal(Rc::new(RefCell::new(self.peek().clone())))
+// All implementation here is private to this module and the effective logic in
+// working with values in the VM.
+impl ValueImpl {
+    fn into_value(self) -> VMResult<Value> {
+        match self {
+            ValueImpl::Invalid => Err(VMStatus::new(StatusCode::INTERNAL_TYPE_ERROR)),
+            ValueImpl::LocalRef(reference) => reference.into_value(),
+            _ => Ok(Value(self)),
+        }
     }
-}
 
-impl Clone for Local {
-    fn clone(&self) -> Self {
+    fn copy_value(&self) -> VMResult<Value> {
         match self {
-            Local::Ref(v) => Local::Ref(v.shallow_clone()),
-            Local::GlobalRef(v) => Local::GlobalRef(v.shallow_clone()),
-            Local::Value(v) => Local::Value(v.clone()),
-            Local::Invalid => Local::Invalid,
+            ValueImpl::Invalid => Err(VMStatus::new(StatusCode::INTERNAL_TYPE_ERROR)),
+            ValueImpl::LocalRef(reference) => reference.copy_value(),
+            _ => Ok(Value(self.clone())),
         }
     }
-}
 
-impl MutVal {
-    pub fn try_own(mv: Self) -> VMResult<Value> {
-        match Rc::try_unwrap(mv.0) {
-            Ok(cell) => Ok(cell.into_inner()),
-            Err(_) => Err(VMStatus::new(StatusCode::LOCAL_REFERENCE_ERROR)),
+    fn borrow_field(&self, field_offset: usize) -> VMResult<Value> {
+        match self {
+            ValueImpl::Struct(s) => s.get_field_reference(field_offset),
+            _ => Err(VMStatus::new(StatusCode::INTERNAL_TYPE_ERROR)),
         }
     }
 
-    pub fn peek(&self) -> Ref<Value> {
-        self.0.borrow()
+    fn size(&self) -> AbstractMemorySize<GasCarrier> {
+        match self {
+            ValueImpl::Invalid | ValueImpl::U64(_) | ValueImpl::Bool(_) => *CONST_SIZE,
+            ValueImpl::Address(_) => AbstractMemorySize::new(ADDRESS_LENGTH as u64),
+            // Possible debate topic: Should we charge based upon the size of the string.
+            // At this moment, we take the view that you should be charged as though you are
+            // copying the string onto the stack here. This doesn't replicate
+            // the semantics that we utilize currently, but this string may
+            // need to be copied at some later time, so we need to charge based
+            // upon the size of the memory that will possibly need to be accessed.
+            ValueImpl::String(s) => words_in(AbstractMemorySize::new(s.len() as u64)),
+            ValueImpl::ByteArray(key) => AbstractMemorySize::new(key.len() as u64),
+            ValueImpl::Struct(s) => s.size(),
+            ValueImpl::Reference(reference) => reference.size(),
+            ValueImpl::GlobalRef(reference) => reference.size(),
+            ValueImpl::LocalRef(reference) => reference.0.size(),
+        }
+    }
+
+    // Structural equality for Move values
+    fn equals(&self, v2: &ValueImpl) -> VMResult<bool> {
```

Comment:
A match on LocalRef is conspicuous by its absence in this function.

Reply:
do you want that to be explicit? not sure I am following the request, if any.
LocalRef should not appear here and it's an error to call equality on them (except in that test case to review as mentioned above). Maybe I should add it for now? thought?